File: server/internal/grammar/reflexive.go
package grammar

import (
	"fmt"
	"strings"
	"unicode"

	"github.com/citavuk/server/internal/lexicon"
)

// Reflexive — возвратная частица «se», относящаяся к разбираемому глаголу.
//
// В сербском «se» почти никогда не стоит вплотную к своему глаголу: это
// клитика, у неё нет своего ударения, и место ей отводит не глагол, а фраза —
// второе место сразу за первым ударным словом. Поэтому «On se zove Marko»
// выглядит так, будто «se» относится к «on», а на самом деле это часть глагола
// «zvati se». Пока разбор показывал одно «zove», читатель видел «зовёт» и не
// понимал, откуда взялось «называется».
type Reflexive struct {
	// Particle — частица как она написана в тексте: «se» или «се».
	Particle string `json:"particle"`
	// Verb — словоформа глагола, к которому частица относится.
	Verb string `json:"verb"`
	// OnParticle — нажали на саму частицу, а не на глагол.
	OnParticle bool `json:"onParticle"`
	// Companion — второе слово пары: то, что нужно подсветить вдобавок к
	// нажатому. Клиент ищет его по написанию и направлению, не пересчитывая
	// байтовые смещения Go в индексы JavaScript.
	Companion string `json:"companion"`
	// Before — спутник стоит ПЕРЕД нажатым словом.
	Before bool `json:"before"`
	// Adjacent — между частицей и глаголом нет других слов.
	Adjacent bool `json:"adjacent"`
	// Phrase — словарный порядок: «zove se», как бы ни стояло в тексте.
	Phrase string `json:"phrase"`
	// Lemma — начальная форма вместе с частицей: «zvati se».
	Lemma string `json:"lemma,omitempty"`
	// Meaning — чем «se» бывает при глаголе.
	Meaning string `json:"meaning"`
	// Why — почему частица оказалась именно на этом месте.
	Why string `json:"why"`
}

// IsVerb сообщает, является ли словоформа глагольной.
type IsVerb func(word string) bool

// sePair находит пару «глагол + частица» вокруг нажатого слова.
//
// Работает в обе стороны. Нажали глагол — ищется частица; нажали саму частицу —
// ищется её глагол. Второе не менее важно: «se» попадается в тексте на каждом
// шагу, и в отрыве от глагола разбирается как местоимение «sebe», а переводится
// и вовсе случайным словом.
//
// Часть предложения ограничивает поиск: в «Vratio se kući, a brat je ostao»
// частица относится к «vratio», и приписывать её «ostao» нельзя.
func sePair(
	sentence string,
	words []seWord,
	index int,
	isVerb IsVerb,
) (verbIndex, seIndex, clauseFrom int, ok bool) {
	from, to := clauseBounds(sentence, words, index)
	if isSeWord(words[index].text) {
		verb := nearestVerb(words, from, to, index, isVerb)
		if verb < 0 {
			return 0, 0, 0, false
		}
		return verb, index, from, true
	}
	se := nearestSe(words, from, to, index)
	if se < 0 {
		return 0, 0, 0, false
	}
	return index, se, from, true
}

func AttachSe(sentence string, start, end int, surface, lemma string, isVerb IsVerb) *Reflexive {
	words := splitWords(sentence)
	index := wordIndexAt(words, start, end)
	if index < 0 {
		return nil
	}
	verbIndex, seIndex, clauseFrom, ok := sePair(sentence, words, index, isVerb)
	if !ok {
		return nil
	}

	particle := words[seIndex].text
	verb := words[verbIndex].text
	onParticle := seIndex == index

	companion, companionIndex := particle, seIndex
	if onParticle {
		companion, companionIndex = verb, verbIndex
	}

	out := &Reflexive{
		Particle:   particle,
		Verb:       verb,
		OnParticle: onParticle,
		Companion:  companion,
		Before:     companionIndex < index,
		Adjacent:   adjacentSe(sentence, words, verbIndex, seIndex),
		Phrase:     verb + " " + strings.ToLower(particle),
		Meaning:    seMeaning(lemma),
		Why:        seWhy(words, clauseFrom, verbIndex, seIndex, verb),
	}
	if lemma != "" {
		out.Lemma = lemma + " " + strings.ToLower(particle)
	}
	return out
}

// SeSpan расширяет границы слова до пары «глагол + se».
//
// Нужно переводчику: «zove» отдельно переводится как «зовёт», а «zove se» — как
// «называется». Одиночное «se» переводится и вовсе случайным словом. Пара
// «глагол + se» — одно слово по смыслу, и разрывать её перед отправкой в
// переводчик значит спрашивать не о том.
//
// Частица через другие слова не захватывается: помеченный фрагмент обязан быть
// сплошным, иначе выравнивание внутри переведённой фразы теряет смысл.
func SeSpan(sentence string, start, end int, isVerb IsVerb) (int, int, bool) {
	words := splitWords(sentence)
	index := wordIndexAt(words, start, end)
	if index < 0 {
		return start, end, false
	}
	verbIndex, seIndex, _, ok := sePair(sentence, words, index, isVerb)
	if !ok || !adjacentSe(sentence, words, verbIndex, seIndex) {
		return start, end, false
	}
	if seIndex < verbIndex {
		return words[seIndex].start, words[verbIndex].end, true
	}
	return words[verbIndex].start, words[seIndex].end, true
}

// SeMeaning объясняет, чем «se» бывает при этом глаголе.
func SeMeaning(lemma string) string { return seMeaning(lemma) }

func seMeaning(lemma string) string {
	if seTantum[lexicon.Normalize(lemma)] {
		return fmt.Sprintf("«%s» без «se» не бывает — это как «-ся» в «бояться».", lemma)
	}
	return "«se» бывает разным: «umiva se» — умывается (сам себя), " +
		"«vide se» — видят друг друга, «ovde se ne puši» — здесь не курят."
}

// seWhy объясняет место частицы: не «так принято», а по какому правилу она
// оказалась здесь. Правило одно на все безударные словечки (se, me, ga, sam,
// je, bih), и, поняв его раз, читатель перестаёт спотыкаться обо все остальные.
//
// Коротко и без терминов. «Клитика» — слово из учебника лингвистики: русскому
// читателю оно ничего не объясняет, а объяснение с ним становится вдвое длиннее
// и перестаёт читаться.
func seWhy(words []seWord, clauseFrom, index, seIndex int, surface string) string {
	position := seIndex - clauseFrom
	switch {
	case position == 0:
		return "«se» — безударная частица и фразу не открывает: ей нужно слово, " +
			"на которое опереться."

	case position == 1 && seIndex-1 == index:
		return fmt.Sprintf(
			"«se» всегда стоит вторым словом. Здесь фразу открывает сам глагол «%s», "+
				"поэтому частица встала сразу за ним.", surface)

	case position == 1:
		return fmt.Sprintf(
			"«se» всегда стоит вторым словом фразы — здесь после «%s». Место ей задаёт "+
				"фраза, а не глагол, поэтому она и оказалась вдали от «%s».",
			words[seIndex-1].text, surface)
	}

	group := make([]string, 0, position)
	for i := clauseFrom; i < seIndex; i++ {
		group = append(group, words[i].text)
	}
	return fmt.Sprintf(
		"«%s» — это одно смысловое начало фразы, и «se» встаёт сразу за ним, "+
			"а не рядом с глаголом «%s».", strings.Join(group, " "), surface)
}

// seWord — слово предложения вместе с байтовыми границами.
type seWord struct {
	text  string
	start int
	end   int
}

func splitWords(sentence string) []seWord {
	out := make([]seWord, 0, 16)
	begin := -1
	for i, r := range sentence {
		if unicode.IsLetter(r) || unicode.IsDigit(r) || r == '-' || r == '\'' {
			if begin < 0 {
				begin = i
			}
			continue
		}
		if begin >= 0 {
			out = append(out, seWord{sentence[begin:i], begin, i})
			begin = -1
		}
	}
	if begin >= 0 {
		out = append(out, seWord{sentence[begin:], begin, len(sentence)})
	}
	return out
}

func wordIndexAt(words []seWord, start, end int) int {
	for i, w := range words {
		if w.start <= start && end <= w.end {
			return i
		}
	}
	return -1
}

// clauseBounds очерчивает часть предложения вокруг слова: диапазон индексов
// [from, to). Границей служит знак препинания либо союз — он относится уже к
// своей части и потому включается в неё слева.
func clauseBounds(sentence string, words []seWord, index int) (int, int) {
	from := index
	for from > 0 {
		if breaksClause(sentence[words[from-1].end:words[from].start]) {
			break
		}
		from--
		if clauseOpeners[lexicon.Normalize(words[from].text)] {
			break
		}
	}

	to := index + 1
	for to < len(words) {
		if breaksClause(sentence[words[to-1].end:words[to].start]) {
			break
		}
		if clauseOpeners[lexicon.Normalize(words[to].text)] {
			break
		}
		to++
	}
	return from, to
}

func breaksClause(gap string) bool {
	return strings.ContainsAny(gap, ".,;:!?()[]{}\"\n\r…«»„“”—–")
}

// nearestSe возвращает индекс ближайшей частицы в пределах части предложения.
// Ближайшей, а не первой: в «Kad se vrati, javi se» обе части свои.
func nearestSe(words []seWord, from, to, index int) int {
	best := -1
	for i := from; i < to; i++ {
		if i == index || !isSeWord(words[i].text) {
			continue
		}
		if best < 0 || distance(i, index) < distance(best, index) {
			best = i
		}
	}
	return best
}

// nearestVerb ищет глагол, к которому относится частица.
//
// Перебор идёт от ближайшего слова наружу: «Bližila se ponoć» — глагол слева,
// «On se zove Marko» — справа, и заранее сторона неизвестна. При равном
// расстоянии первым проверяется слово слева: клитика чаще идёт следом за своим
// глаголом, чем перед ним.
func nearestVerb(words []seWord, from, to, index int, isVerb IsVerb) int {
	if isVerb == nil {
		return -1
	}
	for step := 1; step < to-from; step++ {
		for _, at := range [2]int{index - step, index + step} {
			if at < from || at >= to || at == index {
				continue
			}
			if !isSeWord(words[at].text) && isVerb(words[at].text) {
				return at
			}
		}
	}
	return -1
}

func adjacentSe(sentence string, words []seWord, index, seIndex int) bool {
	if distance(index, seIndex) != 1 {
		return false
	}
	left, right := index, seIndex
	if seIndex < index {
		left, right = seIndex, index
	}
	return strings.TrimSpace(sentence[words[left].end:words[right].start]) == ""
}

func distance(a, b int) int {
	if a > b {
		return a - b
	}
	return b - a
}

func isSeWord(word string) bool {
	return lexicon.Normalize(word) == "se"
}

// clauseOpeners — слова, с которых начинается новая часть предложения. Своя
// клитика есть у каждой части, и без этого списка «Kad se vrati, reći ću mu»
// приписало бы частицу глаголу «reći».
var clauseOpeners = map[string]bool{
	"i": true, "a": true, "ali": true, "pa": true, "te": true, "ili": true,
	"jer": true, "nego": true, "već": true, "da": true, "ako": true,
	"kad": true, "kada": true, "dok": true, "što": true, "iako": true,
	"mada": true, "pošto": true, "ukoliko": true, "čim": true,
	"koji": true, "koja": true, "koje": true, "kojeg": true, "kojem": true,
	"gde": true, "gdje": true, "kako": true, "jel": true,
}

// seTantum — глаголы, у которых «se» входит в словарную форму: без частицы их
// просто нет. Отделять их важно: у остальных «se» — отдельное значение
// (возвратное, взаимное, безличное), и объяснять их одинаково было бы неверно.
var seTantum = map[string]bool{
	"bojati": true, "smejati": true, "smijati": true, "nadati": true,
	"truditi": true, "potruditi": true, "sećati": true, "sjećati": true,
	"setiti": true, "sjetiti": true, "ponašati": true, "dogoditi": true,
	"desiti": true, "diviti": true, "snalaziti": true, "snaći": true,
	"sviđati": true, "svideti": true, "svidjeti": true, "dopadati": true,
	"dopasti": true, "usuditi": true, "rugati": true, "šaliti": true,
	"kajati": true, "boriti": true, "starati": true, "protiviti": true,
	"sastojati": true, "baviti": true, "ticati": true, "čuditi": true,
	"zaljubiti": true, "raspitivati": true, "smilovati": true,
	"nadmetati": true, "prisećati": true, "prisjećati": true,
}

File: frontend/lib/services/reflexive.dart
import '../utils/transliteration.dart';
import 'lexicon_db.dart';

/// Возвратная частица «se» и её глагол.
///
/// В сербском «se» почти никогда не стоит вплотную к своему глаголу: своего
/// ударения у неё нет, и место ей отводит не глагол, а фраза — второе, сразу за
/// первым ударным словом. Поэтому «On se zove Marko» выглядит так, будто «se»
/// относится к «on», а на деле это глагол «zvati se».
///
/// Повторяет server/internal/grammar/reflexive.go: приложение разбирает слово
/// офлайн по своей копии словаря, и разбор обязан совпадать с серверным.
class Reflexive {
  final String particle;
  final String verb;

  /// Нажали на саму частицу, а не на глагол.
  final bool onParticle;

  /// Второе слово пары — то, что нужно подсветить вдобавок к нажатому.
  final String companion;

  /// Спутник стоит ПЕРЕД нажатым словом.
  final bool before;

  /// Между частицей и глаголом нет других слов.
  final bool adjacent;

  /// Словарный порядок: «zove se», как бы ни стояло в тексте.
  final String phrase;

  /// Начальная форма вместе с частицей: «zvati se».
  final String lemma;

  final String meaning;
  final String why;

  const Reflexive({
    required this.particle,
    required this.verb,
    required this.onParticle,
    required this.companion,
    required this.before,
    required this.adjacent,
    required this.phrase,
    required this.lemma,
    required this.meaning,
    required this.why,
  });
}

class _Word {
  final String text;
  final int start;
  final int end;
  const _Word(this.text, this.start, this.end);
}

/// Слова, с которых начинается новая часть предложения.
///
/// Своя частица есть у каждой части, и без этого списка «Kad se vrati, reći ću
/// mu» приписало бы её глаголу «reći».
const _clauseOpeners = <String>{
  'i', 'a', 'ali', 'pa', 'te', 'ili', 'jer', 'nego', 'već', 'da', 'ako',
  'kad', 'kada', 'dok', 'što', 'iako', 'mada', 'pošto', 'ukoliko', 'čim',
  'koji', 'koja', 'koje', 'kojeg', 'kojem', 'gde', 'gdje', 'kako', 'jel',
};

/// Глаголы, у которых «se» входит в словарную форму: без частицы их просто нет.
const _seTantum = <String>{
  'bojati', 'smejati', 'smijati', 'nadati', 'truditi', 'potruditi', 'sećati',
  'sjećati', 'setiti', 'sjetiti', 'ponašati', 'dogoditi', 'desiti', 'diviti',
  'snalaziti', 'snaći', 'sviđati', 'svideti', 'svidjeti', 'dopadati', 'dopasti',
  'usuditi', 'rugati', 'šaliti', 'kajati', 'boriti', 'starati', 'protiviti',
  'sastojati', 'baviti', 'ticati', 'čuditi', 'zaljubiti', 'raspitivati',
  'smilovati', 'nadmetati', 'prisećati', 'prisjećati',
};

String _norm(String s) =>
    SerbianTransliteration.toLatin(s).trim().toLowerCase();

bool _isSe(String word) => _norm(word) == 'se';

final _wordChar = RegExp(r"[\p{L}\p{N}\-']", unicode: true);
final _breaks = RegExp(r'''[.,;:!?()\[\]{}"\n\r…«»„“”—–]''');

List<_Word> _splitWords(String sentence) {
  final out = <_Word>[];
  var begin = -1;
  for (var i = 0; i < sentence.length; i++) {
    if (_wordChar.hasMatch(sentence[i])) {
      if (begin < 0) begin = i;
      continue;
    }
    if (begin >= 0) {
      out.add(_Word(sentence.substring(begin, i), begin, i));
      begin = -1;
    }
  }
  if (begin >= 0) {
    out.add(_Word(sentence.substring(begin), begin, sentence.length));
  }
  return out;
}

int _wordIndexAt(List<_Word> words, int start, int end) {
  for (var i = 0; i < words.length; i++) {
    if (words[i].start <= start && end <= words[i].end) return i;
  }
  return -1;
}

/// Часть предложения вокруг слова. Границей служит знак препинания либо союз —
/// он относится уже к своей части и потому включается в неё слева.
(int, int) _clauseBounds(String sentence, List<_Word> words, int index) {
  var from = index;
  while (from > 0) {
    if (_breaks.hasMatch(sentence.substring(words[from - 1].end, words[from].start))) {
      break;
    }
    from--;
    if (_clauseOpeners.contains(_norm(words[from].text))) break;
  }
  var to = index + 1;
  while (to < words.length) {
    if (_breaks.hasMatch(sentence.substring(words[to - 1].end, words[to].start))) {
      break;
    }
    if (_clauseOpeners.contains(_norm(words[to].text))) break;
    to++;
  }
  return (from, to);
}

int _nearestSe(List<_Word> words, int from, int to, int index) {
  var best = -1;
  for (var i = from; i < to; i++) {
    if (i == index || !_isSe(words[i].text)) continue;
    if (best < 0 || (i - index).abs() < (best - index).abs()) best = i;
  }
  return best;
}

bool _adjacent(String sentence, List<_Word> words, int a, int b) {
  if ((a - b).abs() != 1) return false;
  final left = a < b ? a : b;
  final right = a < b ? b : a;
  return sentence.substring(words[left].end, words[right].start).trim().isEmpty;
}

String _meaning(String lemma) {
  if (_seTantum.contains(_norm(lemma))) {
    return '«$lemma» без «se» не бывает — это как «-ся» в «бояться».';
  }
  return '«se» бывает разным: «umiva se» — умывается (сам себя), '
      '«vide se» — видят друг друга, «ovde se ne puši» — здесь не курят.';
}

/// Объясняет место частицы коротко и без терминов: «клитика» русскому читателю
/// ничего не объясняет, а объяснение с ней становится вдвое длиннее.
String _why(List<_Word> words, int clauseFrom, int verbIndex, int seIndex, String verb) {
  final position = seIndex - clauseFrom;
  if (position == 0) {
    return '«se» — безударная частица и фразу не открывает: ей нужно слово, '
        'на которое опереться.';
  }
  if (position == 1 && seIndex - 1 == verbIndex) {
    return '«se» всегда стоит вторым словом. Здесь фразу открывает сам глагол '
        '«$verb», поэтому частица встала сразу за ним.';
  }
  if (position == 1) {
    return '«se» всегда стоит вторым словом фразы — здесь после '
        '«${words[seIndex - 1].text}». Место ей задаёт фраза, а не глагол, '
        'поэтому она и оказалась вдали от «$verb».';
  }
  final group = [for (var i = clauseFrom; i < seIndex; i++) words[i].text].join(' ');
  return '«$group» — это одно смысловое начало фразы, и «se» встаёт сразу за '
      'ним, а не рядом с глаголом «$verb».';
}

/// Связывает глагол и частицу «se» в пределах одной части предложения.
///
/// Работает в обе стороны: нажали глагол — ищется частица, нажали частицу —
/// её глагол. Второе не менее важно: «se» попадается в тексте на каждом шагу, а
/// в отрыве от глагола разбирается как местоимение «sebe».
Future<Reflexive?> attachSe({
  required String sentence,
  required int start,
  required int end,
  required String surface,
  String lemma = '',
}) async {
  final words = _splitWords(sentence);
  final index = _wordIndexAt(words, start, end);
  if (index < 0) return null;

  Future<String?> verbLemmaOf(String word) => LexiconDb.instance.verbLemma(word);

  final (from, to) = _clauseBounds(sentence, words, index);
  var verbIndex = index;
  int seIndex;
  final onParticle = _isSe(words[index].text);

  if (onParticle) {
    seIndex = index;
    verbIndex = -1;
    // Перебор от ближайшего слова наружу: «Bližila se ponoć» — глагол слева,
    // «On se zove Marko» — справа, и заранее сторона неизвестна.
    for (var step = 1; step < to - from && verbIndex < 0; step++) {
      for (final at in [index - step, index + step]) {
        if (at < from || at >= to || at == index) continue;
        if (_isSe(words[at].text)) continue;
        if (await verbLemmaOf(words[at].text) != null) {
          verbIndex = at;
          break;
        }
      }
    }
    if (verbIndex < 0) return null;
  } else {
    if (await verbLemmaOf(surface) == null) return null;
    seIndex = _nearestSe(words, from, to, index);
    if (seIndex < 0) return null;
  }

  final particle = words[seIndex].text;
  final verb = words[verbIndex].text;
  final verbLemma = lemma.isNotEmpty ? lemma : (await verbLemmaOf(verb) ?? '');
  final companionIndex = onParticle ? verbIndex : seIndex;

  return Reflexive(
    particle: particle,
    verb: verb,
    onParticle: onParticle,
    companion: onParticle ? verb : particle,
    before: companionIndex < index,
    adjacent: _adjacent(sentence, words, verbIndex, seIndex),
    phrase: '$verb ${particle.toLowerCase()}',
    lemma: verbLemma.isEmpty ? '' : '$verbLemma ${particle.toLowerCase()}',
    meaning: _meaning(verbLemma),
    why: _why(words, from, verbIndex, seIndex, verb),
  );
}
